lib/rules/no-constant-binary-expression.js

```diff
@@ -0,0 +1,460 @@
+ * https://262.ecma-international.org/5.1/#sec-11.9.3
+ *
+ * Javascript `==` operator works by converting the boolean to `1` (true) or
+ * `+0` (false) and then checks the values `==` equality to that number.
+ * @param {Scope} scope The scope in which node was found.
+ * @param {ASTNode} node The node to test.
+ * @returns {boolean} Will `node` always coerce to the same boolean value?
+ */
+function hasConstantLooseBooleanComparison(scope, node) {
+    switch (node.type) {
+        case "ObjectExpression":
+        case "ClassExpression":
+
+            /**
+             * In theory objects like:
+             *
+             * `{toString: () => a}`
+             * `{valueOf: () => a}`
+             *
+             * Or a classes like:
+             *
+             * `class { static toString() { return a } }`
+             * `class { static valueOf() { return a } }`
+             *
+             * Are not constant verifiably when `inBooleanPosition` is
+             * false, but it's an edge case we've opted not to handle.
+             */
+            return true;
+        case "ArrayExpression":
+            if (node.elements.length === 1) {
+
+                /*
+                 * Possible future direction if needed: We could check if the
+                 * single value would result in variable boolean comparison.
+                 * For now we will err on the side of caution since `[x]` could
+                 * evaluate to `[0]` or `[1]`.
+                 */
+
+                return false;
+            }
+            return true;
+        case "ArrowFunctionExpression":
+        case "FunctionExpression":
+            return true;
+        case "UnaryExpression":
+            if (node.operator === "void" || // Always returns `undefined`
+                node.operator === "typeof" // All type strings are truthy
+            ) {
+                return true;
+            }
+            if (node.operator === "!") {
+                return isConstant(scope, node.argument, true);
+            }
+
+            /*
+             * We won't try to reason about +, -, ~, or delete
+             * In theory, for the mathematical operators, we could look at the
+             * argument and try to determine if it coerces to a constant numeric
+             * value.
+             */
+            return false;
+        case "NewExpression": // Objects might have custom `.valueOf` or `.toString`.
+            return false;
+        case "CallExpression": {
+            if (node.callee.type === "Identifier" &&
+                node.callee.name === "Boolean" &&
+                isReferenceToGlobalVariable(scope, node.callee)
+            ) {
+                return node.arguments.length === 0 || isConstant(scope, node.arguments[0], true);
+            }
+            return false;
+        }
+        case "Literal": // True or false, literals never change
+            return true;
+        case "Identifier":
+            return node.name === "undefined";
```

Comment:
We had been using `isNullOrUndefined` from `ast-utils.js` which did not perform a scope check. Interestingly, this meant that I was unable to write a regression test for this change because `isNullOr Undefined` would detect the `undefined` side as undefined, and then `hasConstantNullishness` would return true for `true`.

I decided to fork `isNullOrUndefined` and create a local version which does perform the scope check. If we truly consider redefined `undefined` and edge case we need to support, we should consider a followup to add `scope` as an additional argument to `isNullOrUndefined` from `ast-utils.js`.

Reply:
Good idea, we can update `astUtils.isNullOrUndefined` and the rules that are using it in a followup PR, and then switch this rule to it as well.